File: src/main/java/cz/vhromada/catalog/rest/controller/SongController.java
package cz.vhromada.catalog.rest.controller;

import java.util.List;

import cz.vhromada.catalog.entity.Music;
import cz.vhromada.catalog.entity.Song;
import cz.vhromada.catalog.facade.SongFacade;
import cz.vhromada.result.Result;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.util.Assert;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * A class represents controller for songs.
 *
 * @author Vladimir Hromada
 */
@RestController("songController")
@RequestMapping("/catalog/music/{musicId}/songs")
public class SongController extends AbstractCatalogController {

    /**
     * Facade for songs
     */
    private SongFacade songFacade;

    /**
     * Creates a new instance of SongController.
     *
     * @param songFacade facade for songs
     * @throws IllegalArgumentException if facade for songs is null
     */
    @Autowired
    public SongController(final SongFacade songFacade) {
        Assert.notNull(songFacade, "Facade for songs mustn't be null.");

        this.songFacade = songFacade;
    }

    /**
     * Returns song with ID or null if there isn't such song.
     * <br>
     * Validation errors:
     * <ul>
     * <li>ID is null</li>
     * </ul>
     *
     * @param musicId music ID
     * @param songId  song ID
     * @return result with song or validation errors
     */
    @GetMapping("/{songId}")
    public ResponseEntity<Result<Song>> getSong(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId,
        @PathVariable("songId") final Integer songId) {
        return processResult(songFacade.get(songId));
    }

    /**
     * Adds song. Sets new ID and position.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Music is null</li>
     * <li>Music ID is null</li>
     * <li>Music doesn't exist in data storage</li>
     * <li>Song is null</li>
     * <li>Song ID isn't null</li>
     * <li>Song position isn't null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * <li>Length of song is negative value</li>
     * <li>Note is null</li>
     * </ul>
     *
     * @param musicId music ID
     * @param song    song
     * @return result with validation errors
     */
    @PutMapping("/add")
    public ResponseEntity<Result<Void>> add(@PathVariable("musicId") final Integer musicId, @RequestBody final Song song) {
        final Music music = new Music();
        music.setId(musicId);

        return processResult(songFacade.add(music, song), HttpStatus.CREATED);
    }

    /**
     * Updates song.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Song is null</li>
     * <li>ID isn't null</li>
     * <li>Position is null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * <li>Length of song is negative value</li>
     * <li>Note is null</li>
     * <li>Song doesn't exist in data storage</li>
     * </ul>
     *
     * @param musicId music ID
     * @param song    new value of song
     * @return result with validation errors
     */
    @PostMapping("/update")
    public ResponseEntity<Result<Void>> update(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, @RequestBody final Song song) {
        return processResult(songFacade.update(song));
    }

    /**
     * Removes song.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Song is null</li>
     * <li>ID is null</li>
     * <li>Song doesn't exist in song storage</li>
     * </ul>
     *
     * @param musicId music ID
     * @param song    song
     * @return result with validation errors
     */
    @DeleteMapping("/remove")
    public ResponseEntity<Result<Void>> remove(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, @RequestBody final Song song) {
        return processResult(songFacade.remove(song));
    }

    /**
     * Duplicates song.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Song is null</li>
     * <li>ID is null</li>
     * <li>Song doesn't exist in song storage</li>
     * </ul>
     *
     * @param musicId music ID
     * @param song    song
     * @return result with validation errors
     */
    @PostMapping("/duplicate")
    public ResponseEntity<Result<Void>> duplicate(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, @RequestBody final Song song) {
        return processResult(songFacade.duplicate(song), HttpStatus.CREATED);
    }

    /**
     * Moves song in list one position up.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Song is null</li>
     * <li>ID is null</li>
     * <li>Song can't be moved up</li>
     * <li>Song doesn't exist in song storage</li>
     * </ul>
     *
     * @param musicId music ID
     * @param song    song
     * @return result with validation errors
     */
    @PostMapping("/moveUp")
    public ResponseEntity<Result<Void>> moveUp(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, @RequestBody final Song song) {
        return processResult(songFacade.moveUp(song));
    }

    /**
     * Moves song in list one position down.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Song is null</li>
     * <li>ID is null</li>
     * <li>Song can't be moved down</li>
     * <li>Song doesn't exist in song storage</li>
     * </ul>
     *
     * @param musicId music ID
     * @param song    song
     * @return result with validation errors
     */
    @PostMapping("/moveDown")
    public ResponseEntity<Result<Void>> moveDown(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, @RequestBody final Song song) {
        return processResult(songFacade.moveDown(song));
    }

    /**
     * Returns list of songs for specified music.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Music is null</li>
     * <li>ID is null</li>
     * <li>Music doesn't exist in data storage</li>
     * </ul>
     *
     * @param musicId music ID
     * @return result with list of songs or validation error
     */
    @GetMapping({ "", "/list" })
    public ResponseEntity<Result<List<Song>>> findSongsByMusic(@PathVariable("musicId") final Integer musicId) {
        final Music music = new Music();
        music.setId(musicId);

        return processResult(songFacade.find(music));
    }

}

File: src/main/kotlin/cz/vhromada/catalog/rest/controller/SongController.kt
package cz.vhromada.catalog.rest.controller

import cz.vhromada.catalog.entity.Music
import cz.vhromada.catalog.entity.Song
import cz.vhromada.catalog.facade.SongFacade
import cz.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for songs.
 *
 * @author Vladimir Hromada
 */
@RestController("songController")
@RequestMapping("/catalog/music/{musicId}/songs")
class SongController(private val songFacade: SongFacade) : AbstractController() {

    /**
     * Returns song with ID or null if there isn't such song.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *
     * @param musicId music ID
     * @param songId  song ID
     * @return song with ID or null if there isn't such song
     */
    @GetMapping("/{songId}")
    fun getSong(@PathVariable("musicId") musicId: Int,
                @PathVariable("songId") songId: Int): Song? {
        return processResult(songFacade.get(songId))
    }

    /**
     * Adds song. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * Music ID is null
     *  * Music doesn't exist in data storage
     *  * Song ID isn't null
     *  * Song position isn't null
     *  * Name is null
     *  * Name is empty string
     *  * Length of song is negative value
     *  * Note is null
     *
     * @param musicId music ID
     * @param song    song
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@PathVariable("musicId") musicId: Int,
            @RequestBody song: Song) {
        val music = Music(id = musicId, name = null, wikiEn = null, wikiCz = null, mediaCount = null, note = null, position = null)
        processResult(songFacade.add(music, song))
    }

    /**
     * Updates song.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * Length of song is negative value
     *  * Note is null
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param song    new value of song
     */
    @PostMapping("/update")
    fun update(@PathVariable("musicId") musicId: Int,
               @RequestBody song: Song) {
        processResult(songFacade.update(song))
    }

    /**
     * Removes song.
     * <br></br>
     * Validation errors:
     *
     *  * Song doesn't exist in song storage
     *
     * @param musicId music ID
     * @param id      ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("musicId") musicId: Int,
               @PathVariable("id") id: Int) {
        val song = Song(id = id, name = null, length = null, note = null, position = null)
        processResult(songFacade.remove(song))
    }

    /**
     * Duplicates song.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Song doesn't exist in song storage
     *
     * @param musicId music ID
     * @param song    song
     */
    @PostMapping("/duplicate")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@PathVariable("musicId") musicId: Int,
                  @RequestBody song: Song) {
        processResult(songFacade.duplicate(song))
    }

    /**
     * Moves song in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Song can't be moved up
     *  * Song doesn't exist in song storage
     *
     * @param musicId music ID
     * @param song    song
     */
    @PostMapping("/moveUp")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@PathVariable("musicId") musicId: Int,
               @RequestBody song: Song) {
        processResult(songFacade.moveUp(song))
    }

    /**
     * Moves song in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Song can't be moved down
     *  * Song doesn't exist in song storage
     *
     * @param musicId music ID
     * @param song    song
     */
    @PostMapping("/moveDown")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@PathVariable("musicId") musicId: Int,
                 @RequestBody song: Song) {
        processResult(songFacade.moveDown(song))
    }

    /**
     * Returns list of songs for specified music.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Music doesn't exist in data storage
     *
     * @param musicId music ID
     * @return list of songs for specified music
     */
    @GetMapping
    fun findSongsByMusic(@PathVariable("musicId") musicId: Int): List<Song> {
        val music = Music(id = musicId, name = null, wikiEn = null, wikiCz = null, mediaCount = null, note = null, position = null)
        return processResult(songFacade.find(music))!!
    }

}
